### PR TITLE
[CAL-493] Implements tRCP on event types

### PR DESCRIPTION
**Pull request by zomars**
- ~Extracts CreateFirstEventTypeView~
- ~Extracts CreateNewEventDialog~
- Implements tRCP for event types listing

### CODE REVIEW

**Review by emrysal — COMMENTED** [review]
Not many issues I can see, but I need to UI test to be able to approve; and I'm in debug mode on local at the mom so stuff's broken.

File: lib/mutations/event-types/create-event-type.ts [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
-import { CreateEventType } from "@lib/types/event-type";
+import { CreateEventType, CreateEventTypeResponse } from "@lib/types/event-type";
 
 const createEventType = async (data: CreateEventType) => {
-  const response = await fetch.post<CreateEventType, EventType>("/api/availability/eventtype", data);
+  const response = await fetch.post<CreateEventType, CreateEventTypeResponse>(
+    "/api/availability/eventtype",
+    data
+  );
+  console.log(`response`, response);
```
[the remark this answers is not in the document]
Comment (zomars) [reply lib/mutations/event-types/create-event-type.ts:9]: Nice catch

**Review by KATT — CHANGES_REQUESTED** [review]
- Unnecessary "generic" components when they're only used in 1 place
  - I don't think we should functions stuff out of a file before they're used in several places
  - Long files and co-located >> logic split up over N files
  - Don't take my word for it - here's Dan Abramov
    - https://twitter.com/dan_abramov/status/1230841812115501056?lang=en
    - https://twitter.com/dan_abramov/status/1220029098543788032?lang=en
- A button called "Dialog" should not render a visible button
- A Dialog should not handle it's own state (and ideally only the `pages/X`-files deal with things like the URL-params so we don't get collisions later on)
- A component in `components/*` should probably never infer a type from the API - they should all be generic enough to just work without knowledge of API

File: pages/event-types/index.tsx [hunk 2]
```diff
@@ -64,267 +35,59 @@ const EventTypesPage = (props: PageProps) => {
         heading={t("event_types_page_title")}
         subtitle={t("event_types_page_subtitle")}
         CTA={
-          props.eventTypes.length !== 0 && (
-            <CreateNewEventDialog canAddEvents={props.canAddEvents} profiles={props.profiles} />
+          query.data &&
+          query.data.eventTypes.length !== 0 && (
+            <CreateNewEventDialog canAddEvents={query.data.canAddEvents} profiles={query.data.profiles} />
```
[the remark this answers is not in the document]
Comment (emrysal) [reply pages/event-types/index.tsx:40]: You're misunderstanding the Dialog component a little bit yeah - the way it is used here is variable-less. It is in fact a dialog but renders a trigger of that Dialog. When you click that button it triggers.

[the remark this answers is not in the document]
Comment (KATT) [reply pages/event-types/index.tsx:40]: In their demo it's clear that it has a button as a trigger, in our case it's not. However, in our case, the state is controlled by a query param so we should not need to rely on `<DialogTrigger/>` (which I personally feel a bit iffy about as an API)

I.e. (mock code)

```ts

export default function MyPage() {
  const modal = useToggleQuery("new");

  return (
    <>
      {/* Button can be outside - it doesn't matter */}
      <Button href={modal.hrefOn}>Open modal</Button>
      {/* No need for trigger */}
      <Dialog
        open={modal.isOn}
        {/* We should probably add a `modal.setOn(boolean)` or something to get to replace below with `onOpenChange={modal.setOn}` */}
        onOpenChange={(state) => router.replace(state ? modal.hrefOff : modal.hrefOn)}>
        {/* ... */}
      </Dialog>
    </>
  );
}
```

[the remark this answers is not in the document]
Comment (KATT) [reply pages/event-types/index.tsx:40]: Might very well be that I'm misunderstanding something with their modals in general though. And maybe we shouldn't have query param tracked / linkable modals at all.

[the remark this answers is not in the document]
Comment (zomars) [reply pages/event-types/index.tsx:40]: Partially agreed. I don't have much an opinion on this. Just refactored the previous dialog.

File: components/eventtype/CreateNewEventDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+  DropdownMenuItem,
+  DropdownMenuLabel,
+  DropdownMenuSeparator,
+  DropdownMenuTrigger,
+} from "@components/ui/Dropdown";
+import * as RadioArea from "@components/ui/form/radio-area";
+
+type Profiles = inferQueryOutput<"viewer.eventTypes">["profiles"];
+
+const CreateNewEventDialog = ({ profiles, canAddEvents }: { profiles: Profiles; canAddEvents: boolean }) => {
```
[the remark this answers is not in the document]
Comment (zomars) [reply components/eventtype/CreateNewEventDialog.tsx:30]: Agreed except in the vendoring thing. Another point in splitting in files is being able to dynamically load them in the future.

File: components/eventtype/CreateNewEventDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+          {...(canAddEvents
+            ? {
+                href: modalOpen.hrefOn,
+              }
+            : {
+                disabled: true,
+              })}
+          StartIcon={PlusIcon}>
+          {t("new_event_type_btn")}
+        </Button>
```
[the remark this answers is not in the document]
Comment (zomars) [reply components/eventtype/CreateNewEventDialog.tsx:67]: Should we call it Action?

[the remark this answers is not in the document]
Comment (KATT) [reply components/eventtype/CreateNewEventDialog.tsx:67]: It shouldn't contain the state of the button as explained somewhere else here so maybe the CTA could be something like

```tsx
CTA={(
  <>
    <Button ... />
   <Dialog open={modal.isOpen} />
  </>
)}

File: components/eventtype/CreateNewEventDialog.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (zomars) [reply components/eventtype/CreateNewEventDialog.tsx:30]: I consider it prep-work for reducing the initial bundle size using dynamic imports

**Review by KATT — CHANGES_REQUESTED** [review]
> I've been working with React since 2015 so it's more than "a feeling", so trying to put words on my reasoning below. Obviously open for input and to be proven wrong.


## Why colocating code is often preferable over breaking it up into files


### Optimize for code deletability

- Once you remove `pages/X.jsx` and everything is vendored - you'll automatically delete all the dependencies without leaving any traces
- With our linting, you will prompted to delete unused code automatically
- Certainty that you _can_ delete code - deleting something in `components/X` is a lot scarier than `pages/X`

### Context switching 

When one wants to find what component is responsible for a specific thing in the UI - you usually start in the known entry point - in Next.js that is `pages/X.jsx` and navigate to the `export default` - then find the component you're looking for there

Once you then navigate (simple CMD+click) into the component you want to change you'll be faced with two different scenarios:

- **A)** The component is in the same file **and not exported** - you can safely *know* that the component is not re-used and you can modify it without further thinking 
- **B)** The component is `export`ed in a `components/X` directory
  - I now need to at *the minumum*  need to check all places that import this file which breaks my flow of what I'm doing
  - If the component *is reused* I have to consider all of those cases before I change it - changes that also need to be tested.


### "DRY" programming is overrated

Trying to "over-share" code leads to very complicated logic where it's a lot easier to copy and tweak than to make reusable parts for everything. "WET" programming is easy for anyone to pick up and it'll make it a lot cheaper to delete code.

See scenario **B)** on *Context Switching* above


### To answer specific points

> You don't use cmd + P that often I guess. I expect components to live in their respective files for easy finding and navigation. That's the point of React componentization IMO. Files should be kept lean and modular.

No, I don't - I mainly navigate by reading and clicking.


As I pointed out, this has been brought up by React core team members. Below are quoted by Dan Abramov on Twitter:

- > [I still get surprised that “one function per file” is clearly unnecessary but “one component per file” is somehow a common practice!](https://twitter.com/dan_abramov/status/1220029098543788032) 
- > [“One component per file” 🤦‍♂️](https://twitter.com/dan_abramov/status/1230841812115501056?lang=en)



> Good feedback. Also I think we reached the point where abstraction is necessary
> 
> https://kentcdodds.com/blog/when-to-break-up-a-component-into-multiple-components

- 100% for something like a `<Dialog/>` that needs to look and feel consistent but not for something like a `CreateNewEventDialog` which is used once. It's the wrong abstraction - we need to look at the smaller parts - i.e. making a `<FormDialog/>` first
- This blog post is focused on breaking up into multiple _components_ - not multiple _files_
- He also quotes this: [_Duplication is far cheaper than the wrong abstraction. — Sandi Metz_](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction)
- And his ending point is: _So feel free to break up your components into smaller ones, but **don't be afraid of a growing component until you start experiencing real problems**. It's WAY easier to maintain it until it needs to be broken up than maintain a pre-mature abstraction. Good luck!_

## Possible other solutions

- There might be a way to keep components in the `pages/X`-folders without Next.js treating it as a Page - then context switching and the surrounding uncertainty would be reduced
- [Namespaces / Modules](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html)
- Strict folder naming in `components/` to show that code is tied to a specific page. I.e. anything to do with `pages/event-types.tsx` would need to live in somewhere like `components/pages/event-types/MyComponent.tsx`
- Maybe there could be a linter to help us delete "stray" files in `components/` - but since we're [*not really*](https://github.com/calendso/calendso/pull/795) using TypeScript it will be hard to do this with certainty.

**Review by PeerRich — CHANGES_REQUESTED** [review]
merge conflicts

**Comment by zomars** [comment]
Conflicts resolved @PeerRich

File: components/eventtype/CreateNewEventDialog.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (zomars) [reply components/eventtype/CreateNewEventDialog.tsx]: I'll refactor in a following PR

File: components/pages/eventtypes/EventTypeList.tsx [hunk 5]
```diff
@@ -7,25 +7,17 @@ import React, { Fragment } from "react";
...
-    hidden: boolean;
-    id: number;
-    slug: string;
-    title: string;
-    users: {
-      name: string;
-      avatar: string;
-    }[];
-  };
+  types: EventTypes[number]["eventTypes"];
```
Comment (KATT) [inline components/pages/eventtypes/EventTypeList.tsx:20]: Easier to read if you split up the logic a bit

```suggestion
type EventTypeGroup = inferQueryOutput<"viewer.eventTypes">["eventTypes"][number];
type EventType = EventTypeGroup["eventTypes"][number];
interface Props {
  profile: { slug: string | null };
  readOnly: boolean;
  types: EventType[];
}
```

Also when doing this you see that what you return on `<"viewer.eventTypes">["eventTypes"]` should probably be called `<"viewer.eventTypes">["eventTypeGroups"]`

File: server/routers/viewer.tsx [hunk 6]
```diff
@@ -71,6 +71,156 @@ const loggedInViewerRouter = createProtectedRouter()
...
+        }))
+      );
+
+      const canAddEvents = user.plan !== "FREE" || eventTypeGroups[0].eventTypes.length < 1;
+
+      return {
+        canAddEvents,
+        user,
+        // don't display event teams without event types,
+        eventTypes: eventTypeGroups.filter((groupBy) => !!groupBy.eventTypes?.length),
```
Comment (KATT) [inline server/routers/viewer.tsx:214]:
```suggestion
eventTypeGroups: eventTypeGroups.filter((groupBy) => !!groupBy.eventTypes?.length),
```

File: components/pages/eventtypes/EventTypeList.tsx (hunk 5 above)
Comment (KATT) [reply components/pages/eventtypes/EventTypeList.tsx:20]: Also, to echo what I wrote earlier - this is clearly tied to the `viewer.eventTypes` so this component should be inlined in whatever file actually does the `viewer.eventTypes`-useQuery

**Review by KATT — APPROVED** [review]
I don't love the `viewer.eventType`-resolver but I think it's largely unchanged from the `getServerSideProps` so changes to that can be done as a follow-up

File: package.json [hunk 7]
```diff
@@ -15,6 +15,7 @@
     "test": "jest",
     "test-playwright": "jest --config jest.playwright.config.js",
     "test-playwright-lcov": "cross-env PLAYWRIGHT_HEADLESS=1 PLAYWRIGHT_COVERAGE=1 yarn test-playwright && nyc report --reporter=lcov",
+    "type-check": "tsc --pretty --noEmit",
```
[the remark this answers is not in the document]
Comment (zomars) [reply package.json:18]: Isn't causing harm does it? Just some prep work to have that script handy

**Review by PeerRich — APPROVED** [review]

**Pull request merged**